source/opt/loop_fission.cpp

```diff
@@ -0,0 +1,508 @@
+  // |register_threshold_to_split|.
+  split_criteria_ =
+      [register_threshold_to_split](
+          const RegisterLiveness::RegionRegisterLiveness& liveness) {
+        return liveness.used_registers_ > register_threshold_to_split;
+      };
+}
+
+LoopFissionPass::LoopFissionPass() : split_multiple_times_(false) {
+  // Split by default.
+  split_criteria_ = [](const RegisterLiveness::RegionRegisterLiveness&) {
+    return true;
+  };
+}
+
+bool LoopFissionPass::ShouldSplitLoop(const ir::Loop& loop, ir::IRContext* c) {
+  LivenessAnalysis* analysis = c->GetLivenessAnalysis();
+
+  RegisterLiveness::RegionRegisterLiveness liveness{};
+
+  ir::Function* function = loop.GetHeaderBlock()->GetParent();
+  analysis->Get(function)->ComputeLoopRegisterPressure(loop, &liveness);
+
+  return split_criteria_(liveness);
+}
+
+Pass::Status LoopFissionPass::Process(ir::IRContext* c) {
+  bool changed = false;
+
+  for (ir::Function& f : *c->module()) {
+    // We collect all the inner most loops in the function and run the loop
+    // splitting util on each. The reason we do this is to allow us to iterate
+    // over each, as calling creating new loops will invalidate the iterator.
+    std::vector<ir::Loop*> inner_most_loops{};
+    ir::LoopDescriptor& loop_descriptor = *c->GetLoopDescriptor(&f);
+    for (ir::Loop& loop : loop_descriptor) {
+      if (!loop.HasChildren() && ShouldSplitLoop(loop, c)) {
+        inner_most_loops.push_back(&loop);
+      }
+    }
+
+    // List of new loops which meet the criteria to be split again.
+    std::vector<ir::Loop*> new_loops_to_split{};
+
+    while (!inner_most_loops.empty()) {
+      for (ir::Loop* loop : inner_most_loops) {
+        LoopFissionImpl impl{c, loop};
+
+        // Group the instructions in the loop into two different sets of related
+        // instructions. If we can't group the instructions into the two sets
+        // then we can't split the loop any further.
+        if (!impl.GroupInstructionsByUseDef()) {
+          continue;
+        }
+
+        if (impl.CanPerformSplit()) {
+          ir::Loop* second_loop = impl.SplitLoop();
+          changed = true;
+          c->InvalidateAnalysesExceptFor(ir::IRContext::kAnalysisLoopAnalysis);
+
+          // If the newly created loop meets the criteria to be split, split it
+          // again.
+          if (ShouldSplitLoop(*second_loop, c))
+            new_loops_to_split.push_back(second_loop);
+
+          // If the original loop (now split) still meets the criteria to be
+          // split, split it again.
+          if (ShouldSplitLoop(*loop, c)) new_loops_to_split.push_back(loop);
+        }
+      }
+
+      // If the split multiple times flag has been set add the new loops which
+      // meet the splitting criteria into the list of loops to be split on the
+      // next iteration.
+      if (split_multiple_times_) {
+        inner_most_loops = std::move(new_loops_to_split);
```

Comment:
Hmm, doesn't this invalidate the loop iterator?  I'm not actually sure it does.

Reply:
I don't think it should invalidate, since it's just moving a list of pointers to loops managed by the analysis so it's not actually touching anything internal to the loop or analysis. But we aren't iterating directly over the loop at this point anyway, we're iterating over the inner_most_loops list to sidestep the iterator invalidation issue